src/main/java/org/truffleruby/interop/OutgoingForeignCallNode.java

```diff
@@ -99,8 +104,92 @@ protected int objectId(Object receiver, String name, Object[] args,
             limit = "1")
     protected Object readArrayElement(Object receiver, String name, Object[] args,
             @Cached(value = "name", allowUncached = true) @Shared("name") String cachedName,
+            @CachedLibrary("receiver") InteropLibrary interop,
+            @Cached TranslateInteropExceptionNode translateInteropException,
+            @CachedContext(RubyLanguage.class) RubyContext context,
             @Cached InteropNodes.ReadArrayElementNode readNode) {
-        return readNode.execute(receiver, args[0]);
+        return at(receiver, name, args, cachedName, interop, translateInteropException, context, readNode);
+    }
+
+    @Specialization(
+            guards = {
+                    "name == cachedName",
+                    "cachedName.equals(AT)",
+                    "args.length == 1",
+                    "isBasicInteger(first(args))" },
+            limit = "1")
+    protected Object at(Object receiver, String name, Object[] args,
+            @Cached(value = "name", allowUncached = true) @Shared("name") String cachedName,
+            @CachedLibrary("receiver") InteropLibrary interop,
+            @Cached TranslateInteropExceptionNode translateInteropException,
+            @CachedContext(RubyLanguage.class) RubyContext context,
+            @Cached InteropNodes.ReadArrayElementNode readNode) {
+        try {
+            long args0 = (LongCastNodeGen.create()).executeCastLong(args[0]);
+            long size = interop.getArraySize(receiver);
+            if (args0 < -size || args0 >= size) {
+                return nil;
+            }
+            return (args0 < 0)
+                    ? readNode.execute(receiver, size + args0)
+                    : readNode.execute(receiver, args0);
```

Comment:
Take a look at how index normalization is done in org.truffleruby.core.array.ArrayNodes.IndexNode.
Both conditions should use a ConditionProfile, and `if < 0; += size` should be done first to simplify the logic.
Also, we should not have two calls to `readNode.execute()` but a single one.